src/vm/checker/typecheck/mod.rs

```diff
@@ -0,0 +1,404 @@
+                    // TODO: these inserts need to check if the name already exists!
+                    match function_name.as_str() {
+                        "define" => {
+                            if function_args.len() < 1 {
+                                return Err(CheckError::new(CheckErrors::DefineFunctionBadSignature))
+                            } else {
+                                if function_args[0].match_list().is_some() {
+                                    let (f_name, f_type) = self.type_check_define_function(expression,
+                                                                                           context)?;
+                                    self.contract_context.function_types.insert(f_name, f_type);
+                                    Ok(Some(()))
+                                } else {
+                                    let (v_name, v_type) = self.type_check_define_variable(function_args,
+                                                                                           context)?;
+                                    self.contract_context.variable_types.insert(v_name, v_type);
+                                    Ok(Some(()))
+                                }
+                            }
+                        },
+                        "define-public" => {
+                            let (f_name, f_type) = self.type_check_define_function(expression,
+                                                                                   context)?;
+                            if !TypeSignature::new_atom(AtomTypeIdentifier::BoolType).admits_type(
+                                &f_type.return_type()) {
+                                Err(CheckError::new(CheckErrors::PublicFunctionMustReturnBool))
+                            } else {
+                                self.contract_context.public_function_types.insert(f_name, f_type);
+                                Ok(Some(()))
+                            }
+                        },
+                        "define-map" => {
+                            let (f_name, f_type) = self.type_check_define_map(expression,
+                                                                              context)?;
+                            self.contract_context.map_types.insert(f_name, f_type);
+                            Ok(Some(()))
+                        },
+                        _ => {
+                            Ok(None)
+                        }
+                    }
+                } else {
+                    Ok(None)
+                }
+            } else {
+                Ok(None) // not a define
+            }
+        } else {
+            Ok(None) // not a define.
+        }
+    }
+}
+
+pub fn type_check_contract(contract: &mut [SymbolicExpression], analysis_db: &AnalysisDatabase) -> CheckResult<ContractAnalysis> {
+    let mut type_checker = TypeChecker::new(analysis_db);
```

Comment:
Not that this is a bug, but in keeping with RAII, do you think it would be more appropriate to alter the constructor for `TypeChecker` to instead look like:

```
let mut contract_context = TypingContext::new();
let mut type_checker = TypeChecker::from_contract(contract, analysis_db, &mut contract_context)?;
```

The `from_contract` method would replace the `for exp in contract` loop body, so a `TypeChecker` could only be instantiated if it was able to evaluate the contract.  This way, we effectively hide all the mutations that `type_check()` does from the caller, and make it impossible for those mutations to be partially observed or accidentally applied by the caller.  What do you think?

Reply:
Yeah, I think this is closer to "good" than the current implementation. However, the intent of the TypeChecker is really only ever to be invoked on one contract (it's something like a visitor used for a single pass), so it might make more sense if it had a private constructor, and this function, type_check_contract was a static method of TypeChecker.

Reply:
Refactored to be a static method.